sw/host/opentitantool/opentitantool_derive/src/lib.rs

```diff
@@ -0,0 +1,55 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, Data, DataEnum, DeriveInput, Fields, Ident, Variant};
+
+/// Derives the `CommandDispatch` trait for a NewType enum.
+/// 
+/// Derive this on purely interior nodes in the opentitantool command hierarchy
+/// (that is: nodes whose only purpose is to dispatch down to the next layer
+/// in the command heirarchy).
+#[proc_macro_derive(CommandDispatch)]
+pub fn derive_command_dispatch(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let input = parse_macro_input!(input as DeriveInput);
+
+    let expanded = match &input.data {
+        Data::Enum(e) => dispatch_enum(&input.ident, e),
+        _ => unimplemented!("CommandDispatch is only implemented for enums"),
+    };
+    proc_macro::TokenStream::from(expanded)
+}
+
+fn dispatch_enum(name: &Ident, e: &DataEnum) -> TokenStream {
+    let arms = e
+        .variants
+        .iter()
+        .map(|variant| dispatch_variant(name, variant))
+        .collect::<Vec<_>>();
+
+    quote! {
+        impl crate::command::CommandDispatch for #name {
+            fn run(&self, backend: &mut dyn opentitanlib::transport::Transport) -> anyhow::Result<Option<Box<dyn erased_serde::Serialize>>> {
+                match self {
+                    #(#arms),*
+                }
+            }
+        }
+    }
+}
+
+fn dispatch_variant(name: &Ident, variant: &Variant) -> TokenStream {
+    let ident = &variant.ident;
+    let unnamed = match &variant.fields {
+        Fields::Unnamed(u) => u.unnamed.iter().collect::<Vec<_>>(),
+        _ => unimplemented!("CommandDispatch is only implemented for Newtype variants"),
+    };
+    if unnamed.len() != 1 {
+        unimplemented!("CommandDispatch is only implemented for Newtype variants");
+    }
+    quote! {
+        #name::#ident(ref __field) => __field.run(backend)
```

Comment:
Perhaps it may be useful to use a fully-qualified method call, so that you call the right trait's method?

Reply:
Done.